### ISSUE TITLE
cannot set container ulimits from lxc configuration

### ISSUE THREAD

**Issue by kfix** [lxc/lxc#1175]
I have some unprivileged containers that run software which opens large amounts of file handles.
Currently, there is no lxc config option to allow pre-setting a container's ulimits at launch, so these processes will be killed after they try to setrlimit() their desired ulimits, fail, and then use up too many file handles.

As a workaround, I launch these containers and then manually:
- find the container's /sbin/init PID in the parent namespace (say, 12345)
- `prlimit --pid=12345 --nofile=1000000` to allow container's processes to `ulimit -n 100000`
- `systemctl restart hungry-software`

Obviously that's not scalable. A hook could be made to do this, but that's ugly and hooks are not supported in LXD either.

I found this proposal to implement a config flag to allow setting ulimits before spawning container init:
https://lists.linuxcontainers.org/pipermail/lxc-users/2015-September/010091.html

I'm in no position to do this work myself but am raising this issue for tracking.

**Comment by kfix** [comment]
As an example, I've configured lxd 2.2 on ubuntu 16.04 to have a larger number of NOFILE:

```
<520> $ cat /etc/systemd/system/lxd.service.d/nofiles.conf
[Service]
LimitNOFILE=1048576
LimitNOFILESoft=1048576

<521> $ sudo systemctl status lxd
● lxd.service - LXD - main daemon
   Loaded: loaded (/lib/systemd/system/lxd.service; indirect; vendor preset: enabled)
  Drop-In: /etc/systemd/system/lxd.service.d
           └─nofiles.conf
   Active: active (running) since Fri 2016-09-16 18:51:06 PDT; 8min ago
     Docs: man:lxd(1)
  Process: 29711 ExecStartPost=/usr/bin/lxd waitready --timeout=600 (code=exited, status=0/SUCCESS)
  Process: 29708 ExecStartPost=/usr/lib/lxd/profile-config (code=exited, status=0/SUCCESS)
  Process: 29695 ExecStartPre=/usr/lib/x86_64-linux-gnu/lxc/lxc-apparmor-load (code=exited, status=0/SUCCESS)
 Main PID: 29707 (lxd)
    Tasks: 18
   Memory: 13.9M
      CPU: 438ms
   CGroup: /system.slice/lxd.service
           └─29707 /usr/bin/lxd --group lxd --logfile=/var/log/lxd/lxd.log

Sep 16 18:51:06 krampus systemd[1]: Starting LXD - main daemon...
Sep 16 18:51:06 krampus lxd[29707]: t=2016-09-16T18:51:06-0700 lvl=warn msg="CGroup memory swap accounting is disabled, swap limits will be ignored."
Sep 16 18:51:06 krampus systemd[1]: Started LXD - main daemon.

<522> $ sudo /usr/bin/prlimit -p 29707
RESOURCE   DESCRIPTION                             SOFT      HARD UNITS
AS         address space limit                unlimited unlimited bytes
CORE       max core file size                         0 unlimited blocks
CPU        CPU time                           unlimited unlimited seconds
DATA       max data size                      unlimited unlimited bytes
FSIZE      max file size                      unlimited unlimited blocks
LOCKS      max number of file locks held      unlimited unlimited
MEMLOCK    max locked-in-memory address space     65536     65536 bytes
MSGQUEUE   max bytes in POSIX mqueues            819200    819200 bytes
NICE       max nice prio allowed to raise             0         0
NOFILE     max number of open files             1048576   1048576
NPROC      max number of processes            unlimited unlimited
RSS        max resident set size              unlimited unlimited pages
RTPRIO     max real-time priority                     0         0
RTTIME     timeout for real-time tasks        unlimited unlimited microsecs
SIGPENDING max number of pending signals         256666    256666
STACK      max stack size                       8388608 unlimited bytes
```

However, any unprivileged containers I start will have a "default" NOFILE setting:
```
<523> $ lxc start joey-lxd
<524> $ sudo /usr/bin/prlimit -p $(lxc info joey-lxd | awk '$1=="Pid:"{print $2}')
RESOURCE   DESCRIPTION                             SOFT      HARD UNITS
AS         address space limit                unlimited unlimited bytes
CORE       max core file size                         0 unlimited blocks
CPU        CPU time                           unlimited unlimited seconds
DATA       max data size                      unlimited unlimited bytes
FSIZE      max file size                      unlimited unlimited blocks
LOCKS      max number of file locks held      unlimited unlimited
MEMLOCK    max locked-in-memory address space     65536     65536 bytes
MSGQUEUE   max bytes in POSIX mqueues            819200    819200 bytes
NICE       max nice prio allowed to raise             0         0
NOFILE     max number of open files               65536     65536
NPROC      max number of processes            unlimited unlimited
RSS        max resident set size              unlimited unlimited pages
RTPRIO     max real-time priority                     0         0
RTTIME     timeout for real-time tasks        unlimited unlimited microsecs
SIGPENDING max number of pending signals         256666    256666
STACK      max stack size                       8388608 unlimited bytes
```

the container cannot increase its own limit, so I have to tweak it from outside:
```
<525> $ sudo /usr/bin/prlimit -p $(lxc info joey-lxd | awk '$1=="Pid:"{print $2}') --nofile=1048577:1048577
prlimit: failed to set the NOFILE resource limit: Operation not permitted
<526> $ sudo /usr/bin/prlimit -p $(lxc info joey-lxd | awk '$1=="Pid:"{print $2}') --nofile=1048576:1048576
<527> $
```

now the hungry processes inside joey-lxd (also in nested containers) can self-set the ulimits they need.

**Comment by kfix** [comment]
I'm using the following script from my LXD physical-host's `/etc/rc.local` to work around this:

```
#!/bin/bash -x
# raise up an unprivileged container's NOFILE hard-limit
PATH=$PATH:/usr/bin
ctr=$1
maxfds=$(sysctl -n fs.nr_open)
lxc info # kicks off lxd startup...
while (! lxc exec $ctr -- true); do sleep 5; printf .; done # wait $ctr startup
initpid=$(lxc info $ctr | awk '$1=="Pid:"{print $2}')
prlimit -p $initpid --nofile=:$maxfds
# now find the lxd daemon in the container and reset its limits
#   this allows it to launch grand-child containers with file-handle hungry processes
lxc exec $ctr -- lxc info # wait for lxd-in-ctr startup
lxdpid=$(pgrep -P $initpid -o lxd)
[ -n "$lxdpid" ] && prlimit -p $lxdpid --nofile=:$maxfds
```

**Comment by brauner** [comment]
Closed by #1276.
